FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -7,36 +7,75 @@ import CoreData
 public final class CoreDataFeedStore: FeedStore {
 	private static let modelName = "FeedStore"
 	private static let model = NSManagedObjectModel(name: modelName, in: Bundle(for: CoreDataFeedStore.self))
-
+	
 	private let container: NSPersistentContainer
 	private let context: NSManagedObjectContext
-
+	
 	struct ModelNotFound: Error {
 		let modelName: String
 	}
-
+	
 	public init(storeURL: URL) throws {
 		guard let model = CoreDataFeedStore.model else {
 			throw ModelNotFound(modelName: CoreDataFeedStore.modelName)
 		}
-
+		
 		container = try NSPersistentContainer.load(
 			name: CoreDataFeedStore.modelName,
 			model: model,
 			url: storeURL
 		)
 		context = container.newBackgroundContext()
 	}
-
+	
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let coreDataFeed = try CoreDataFeed.fetch(in: context) {
+					completion(.found(feed: coreDataFeed.localFeed, timestamp: coreDataFeed.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
-
+	
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let newCache = try CoreDataFeed.newInstance(in: context)
+				
+				let value = CoreDataFeed.managedFeedSet(from: feed, in: context)
+				
+				newCache.coreDataFeedImages = value
+				newCache.timestamp = timestamp
+				
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
-
+	
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try CoreDataFeed.fetch(in: context).map(context.delete)
+				try context.save()
```

Comment:
In case there is no previously stored feed, `context.delete` will not change the context, thus there will be no changes in it and will be no need to save it. Otherwise, Core Data may perform unnecessary work.

Please save the context only in case the feed was deleted.

```suggestion
				if let feed = try CoreDataFeed.fetch(in: context) {
					context.delete(feed)
					try context.save()
				}
```

Reply:
Done